lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -1617,6 +1618,191 @@ FullApplySite SILCombiner::rewriteApplyCallee(FullApplySite apply,
   }
 }
 
+bool SILCombiner::tyrOptimizeStringSwitchIn(SILFunction *F) {
+  SILModule &module = F->getModule();
+
+  // Find all the string literals used in a series of comparisons agains
+  // one "normal" string.
+  SmallVector<StringLiteralInst *, 12> stringLiterals;
+  SmallVector<ApplyInst *, 12> stringCompares;
+  SILValue subject;
+  for (auto &block : F->getBlocks()) {
+    for (auto &inst : block) {
+      auto *apply = dyn_cast<ApplyInst>(&inst);
+      if (!apply || !apply->getReferencedFunctionOrNull())
+        continue;
+      // Filter out the string.equals functions.
+      if (!apply->getReferencedFunctionOrNull()->hasSemanticsAttr(
+              "string.equals"))
+        continue;
+
+      // Make sure that we are comparing agains the same subject as the
+      // other comparisons.
+      bool argsEqual =
+          apply->getArgument(1) != subject && apply->getArgument(0) != subject;
+
+      if (!stringLiterals.empty() &&
+          (argsEqual || !DA->get(F)->dominates(stringLiterals.front(), apply)))
```

Comment:
👍 for the dominance check.
But this still relies on the fact that the "first" (i.e. most dominant) check is the first one in the function's block list. The order of basic blocks in the function can be arbitrary.
You could do something like:
```
if (commonDominator dominates apply) {
  // ok
} else if (apply dominates commonDominator) {
  commonDominator = apply
} else {
  bail
}
```

Reply:
Hmm. I don't quite understand. I don't think the situation you're talking about would ever be able to exist in a valid program. If I understand correctly, the issue is that we might bail on a function where the comparison (`apply`) is dominated by the first string literal we collected, but that string literal doesn't have anything to do with `apply`. Is that correct? We collect the comparisons in order, meaning `apply` is _always_ dominated by `stringCompares.back()`. The comparison is only valid if `apply`'s literal argument dominates `apply`. Therefore, the first string literal must dominate all string comparisons we care about. That's my logic anyway. There might be an issue that I do not see, though.

Reply:
I think you misunderstood (maybe my pseudo code was too "pseudo").
All I'm saying is that it would be nice if this algorithm would work without relying on the basic block ordering in the function. Basic blocks can be ordered like that, e.g.
```
bb0:
   ...
  br bb2
bb1:
  ...
  return
bb2:
  ...
  br bb1
```
if you do "for (auto &block : F->getBlocks()) {" there is no guarantee that the blocks are in any dominance or whatever order.
So you might hit the "second" comparison before the "first" comparison.

Reply:
@eeckstein would iterating through a `DominanceOrder` resolve this? 

I'll add a .sil test to make sure it gets resolved.

Reply:
Yeah, using DominanceOrder would be a good idea

Reply:
Sorry for the delay. This pass now iterates over the blocks in dominance order. I also moved the pass into a function called `doOneIteration` so that it runs the pass until there is nothing left to do. This way multiple string comparisons get optimized in the same pass.